# Car Dataset Exploratory Data Analysis

## Objectives:

- Understand dataset structure
- Detect data quality issues
- Clean the data
- Analyze relationships
- Generate insights

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv("car_price_dataset.csv")

## Information:

In [3]:
pd.set_option('display.max_columns', None)
df.head(20)

,Brand,Model,Year,CarAge,Condition,Mileage(km),EngineSize(L),FuelType,Horsepower,Torque,Transmission,DriveType,BodyType,Doors,Seats,Color,Interior,Options,City,AccidentHistory,Insurance,RegistrationStatus,FuelEfficiency(L/100km),PricePerKm,Price($)
0,Porsche,Panamera,2008,17,Used,256395,3.3,Gasoline,513,395,Manual,AWD,Sedan,4,4,White,Cloth,"Navigation, Cruise Control, Heated Seats, Blue...",Tehran,No,Valid,Incomplete,11.96,0.05,13884
1,Audi,A6,2023,2,Used,20433,2.2,Diesel,302,270,Manual,FWD,Sedan,4,5,Black,Cloth,"Parking Sensors, Cruise Control, Touchscreen",Berlin,Yes,Expired,Incomplete,8.74,1.90,38888
2,BMW,X5,2022,3,Used,52328,3.2,Gasoline,400,388,Automatic,AWD,SUV,5,7,Gray,Leather,"Touchscreen, Bluetooth, Cruise Control, Naviga...",Tokyo,Yes,Valid,Complete,15.68,0.63,33074
3,Hyundai,Tucson,2019,6,Used,91878,1.6,Hybrid,187,219,Automatic,FWD,SUV,5,5,Silver,Cloth,"Sunroof, Rear Camera, Bluetooth, Parking Senso...",Delhi,No,Expired,Complete,9.45,0.14,12966
4,Fiat,500,2012,13,Damaged,192331,1.1,Gasoline,90,112,Automatic,FWD,Hatchback,3,4,Red,Leather,"Heated Seats, Touchscreen",Delhi,No,Valid,Complete,7.16,0.01,2670
5,Porsche,911 Carrera,2018,7,Used,110968,3.3,Gasoline,565,392,Automatic,AWD,Coupe,2,2,White,Cloth,Sunroof,Paris,No,Expired,Incomplete,12.69,0.43,47830
6,Mercedes-Benz,S-Class,2019,6,Used,82607,5.8,Hybrid,509,750,Manual,RWD,Sedan,4,5,Silver,Cloth,"Navigation, Parking Sensors, Touchscreen",Los Angeles,Yes,Expired,Complete,17.83,0.62,51189
7,Porsche,Panamera,2014,11,Used,163074,3.5,Gasoline,513,428,Manual,RWD,Sedan,4,4,Red,Cloth,"Bluetooth, Touchscreen, Cruise Control, Heated...",Tokyo,No,Valid,Complete,13.06,0.17,27296
8,Audi,Q7,2007,18,Used,274471,2.0,Gasoline,315,277,Automatic,AWD,SUV,5,7,Red,Leather,"Touchscreen, Bluetooth",Tokyo,Yes,Expired,Incomplete,10.38,0.02,5389
9,Ford,Mustang,2019,6,Used,95498,4.6,Gasoline,364,588,Manual,RWD,Coupe,2,2,Red,Cloth,"Touchscreen, Cruise Control, Navigation, Bluet...",Berlin,Yes,Valid,Complete,14.90,0.11,10704


In [4]:
df.shape

(50000, 25)

In [5]:
df.dtypes

Brand                          str
Model                          str
Year                         int64
CarAge                       int64
Condition                      str
Mileage(km)                  int64
EngineSize(L)              float64
FuelType                       str
Horsepower                   int64
Torque                       int64
Transmission                   str
DriveType                      str
BodyType                       str
Doors                        int64
Seats                        int64
Color                          str
Interior                       str
Options                        str
City                           str
AccidentHistory                str
Insurance                      str
RegistrationStatus             str
FuelEfficiency(L/100km)    float64
PricePerKm                 float64
Price($)                     int64
dtype: object

In [6]:
df.isnull().sum()

Brand                         0
Model                         0
Year                          0
CarAge                        0
Condition                     0
Mileage(km)                   0
EngineSize(L)                 0
FuelType                      0
Horsepower                    0
Torque                        0
Transmission                  0
DriveType                     0
BodyType                      0
Doors                         0
Seats                         0
Color                         0
Interior                      0
Options                       0
City                          0
AccidentHistory               0
Insurance                     0
RegistrationStatus            0
FuelEfficiency(L/100km)       0
PricePerKm                 2459
Price($)                      0
dtype: int64

In [7]:
missing=df.isnull().mean()*100
print(missing.sort_values(ascending=False))

PricePerKm                 4.918
Model                      0.000
Year                       0.000
CarAge                     0.000
Condition                  0.000
Mileage(km)                0.000
EngineSize(L)              0.000
FuelType                   0.000
Brand                      0.000
Horsepower                 0.000
Torque                     0.000
DriveType                  0.000
Transmission               0.000
Doors                      0.000
Seats                      0.000
Color                      0.000
BodyType                   0.000
Interior                   0.000
Options                    0.000
AccidentHistory            0.000
City                       0.000
Insurance                  0.000
RegistrationStatus         0.000
FuelEfficiency(L/100km)    0.000
Price($)                   0.000
dtype: float64


## Cleaning:

### Making all the columns lower case and removing extra spaces

In [8]:
df.columns=(df.columns.str.lower().str.strip())

### Renaming of the columns

In [9]:
df = df.rename(columns={"enginesize(l)": "enginesize"})
df = df.rename(columns={"fuelefficiency(l/100km)": "fuelefficiency"})
df = df.rename(columns={"price($)": "price"})
df = df.rename(columns={"mileage(km)": "mileage"})

### Defining Scope
- We are removing rows with missing priceperkm because we only want to deal with old cars.
- Since out Of 2,459 missing priceperkm values, 2,454 belong to New cars and only 5 belong to 
- Used cars that is why we are dropping these rows which will remove all New cars (out of scope) 

In [10]:
df['condition'].value_counts()

condition
Used       42767
Damaged     4779
New         2454
Name: count, dtype: int64

In [11]:
df=df.dropna(subset="priceperkm")

In [17]:
df['condition'].value_counts()

condition
Used       42762
Damaged     4779
Name: count, dtype: int64

### Mean, Median and other important calculations

In [12]:
df.describe()

,year,carage,mileage,enginesize,horsepower,torque,doors,seats,fuelefficiency,priceperkm,price
count,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000,47541.000000
mean,2014.481311,10.518689,157783.089670,2.126249,279.776740,326.720978,4.139311,4.722955,9.343063,0.305946,14462.220904
std,5.766638,5.766638,86655.715546,1.139281,153.655567,136.972269,0.978764,1.080637,4.076293,0.855643,14930.163047
min,2005.000000,1.000000,1081.000000,0.000000,65.000000,16.000000,2.000000,2.000000,0.000000,0.000000,1249.000000
25%,2010.000000,6.000000,82792.000000,1.400000,179.000000,227.000000,4.000000,5.000000,7.530000,0.020000,4729.000000
50%,2014.000000,11.000000,158571.000000,2.000000,240.000000,307.000000,4.000000,5.000000,9.680000,0.060000,9468.000000
75%,2019.000000,15.000000,232587.000000,2.800000,386.000000,415.000000,5.000000,5.000000,11.980000,0.220000,18177.000000
max,2024.000000,20.000000,320262.000000,6.000000,905.000000,850.000000,5.000000,7.000000,21.180000,34.560000,117598.000000


### Using mapping dictionary to change simple binary values into booleans

In [13]:
bool_map={'Yes':True,'No':False}
bool_map2={'Valid':True,'Expired':False}
bool_map3={'Incomplete':True,'Complete':False}
df["accidenthistory"]=df["accidenthistory"].map(bool_map)
df["insurance"]=df["insurance"].map(bool_map2)
df["registrationstatus"]=df["registrationstatus"].map(bool_map3)

## Final Cleansed Data

In [14]:
df.head()

,brand,model,year,carage,condition,mileage,enginesize,fueltype,horsepower,torque,transmission,drivetype,bodytype,doors,seats,color,interior,options,city,accidenthistory,insurance,registrationstatus,fuelefficiency,priceperkm,price
0,Porsche,Panamera,2008,17,Used,256395,3.3,Gasoline,513,395,Manual,AWD,Sedan,4,4,White,Cloth,"Navigation, Cruise Control, Heated Seats, Blue...",Tehran,False,True,True,11.96,0.05,13884
1,Audi,A6,2023,2,Used,20433,2.2,Diesel,302,270,Manual,FWD,Sedan,4,5,Black,Cloth,"Parking Sensors, Cruise Control, Touchscreen",Berlin,True,False,True,8.74,1.90,38888
2,BMW,X5,2022,3,Used,52328,3.2,Gasoline,400,388,Automatic,AWD,SUV,5,7,Gray,Leather,"Touchscreen, Bluetooth, Cruise Control, Naviga...",Tokyo,True,True,False,15.68,0.63,33074
3,Hyundai,Tucson,2019,6,Used,91878,1.6,Hybrid,187,219,Automatic,FWD,SUV,5,5,Silver,Cloth,"Sunroof, Rear Camera, Bluetooth, Parking Senso...",Delhi,False,False,False,9.45,0.14,12966
4,Fiat,500,2012,13,Damaged,192331,1.1,Gasoline,90,112,Automatic,FWD,Hatchback,3,4,Red,Leather,"Heated Seats, Touchscreen",Delhi,False,True,False,7.16,0.01,2670


In [15]:
df.shape

(47541, 25)

### Note: we left options column untouched but we may clean it in future

In [16]:
df_check = pd.read_csv("car_price_dataset.csv")
df_check.columns = df_check.columns.str.lower().str.strip()
df_check[df_check["priceperkm"].isnull()]["condition"].value_counts()

condition
New     2454
Used       5
Name: count, dtype: int64